### PR TITLE
Add support for serving on Unix domain socket

### PR DESCRIPTION
**Pull request by hsartoris-bard**
## Summary
These changes take advantage of `jetty-unixsocket` to allow serving on a socket, as in the following example:

```clojure
(run-jetty handler {:http? false, :socket "/run/someapp.sock"})
```
```bash
## hostname is still required for making requests
curl --unix-socket /run/someapp.sock http://my-expected-hostname/url-path
```

## Rationale
Unix sockets are simpler to implement access controls to than TCP sockets, as you can simply use file permissions. They are supported by common reverse proxies such as Apache and NGINX, and are thus reasonable for exposing a service through a proxy that may be handling authentication, without making that service available to all users on the host machine.

## Possible enhancements:
1. Options for file ownership and access controls
2. A secure-by-default approach might see `:http?` default false when `:socket` is set, unless `:host` or `:port` are explicitly configured.
3. To avoid the additional dependency, socket support could be conditional on manually including the `jetty-unixsocket` library. It is not an overwhelmingly heavy dependency, though.

Note: it has proved nontrivial to get a test working, as clj-http does not support sockets, and the Clojure implementations that I've found do not seem to work as intended out of the box. I have verified that this implementation is operational manually, and I am willing to work through the process of getting a test working as long as you're interested in incorporating this functionality.

### CODE REVIEW

**Comment by weavejester** [comment]
I think this is reasonable in theory. Perhaps `:unix-socket` instead of `:socket` as an option, so it's not confused with other types of sockets. Also, have you tested to ensure that the inclusion of this library doesn't cause issues when using it under operating systems that don't have sockets (e.g. Windows)? I'd also suggest not automatically deleting the socket file on startup - that feels potentially dangerous. Instead, raise an error.

**Comment by hsartoris-bard** [comment]
I was under the impression that there was now unix socket support on Windows, but indeed it appears not to work. 

## Changes

- [X] check for windows
- [X] rename `:socket` to `:unix-socket`
- [X] make `:http?` default false when `:unix-socket` is set, except when `:host` or `:port` are set
- [X] throw error on extant file instead of deleting
- [X] tests for basic functionality and errant attempts to use extant file

Let me know if there's anything else you'd like to see!

**Comment by hsartoris-bard** [comment]
Whoops, wasn't my intention to check in the `*warn-on-reflection*` line - I can remove that, or resolve the singular reflection warning.

File: ring-jetty-adapter/src/ring/adapter/jetty.clj [hunk 1]
```diff
@@ -130,6 +133,25 @@
...
+                     :unix-socket unix-socket})))
+  (let [http-factory (HttpConnectionFactory. (http-config options))
+        socket (io/file unix-socket)]
+    (when (.exists socket)
+      (throw (ex-info "File already exists at socket path; should be deleted before starting server"
+                      {:unix-socket unix-socket})))
+    (.deleteOnExit socket)
+    (doto (UnixSocketConnector.
+            server
+            #^"[Lorg.eclipse.jetty.server.ConnectionFactory;" (into-array ConnectionFactory [http-factory]))
```
Comment (weavejester) [inline ring-jetty-adapter/src/ring/adapter/jetty.clj:151]: Can you split the line over 80 characters into two lines? Also can you fix the indentation so it's according to the style guide:

```clojure
(doto (UnixSocketConnector.
       server
       ^"[Lorg.eclipse.jetty.server.ConnectionFactory;"
       (into-array ConnectionFactory [http-factory]))
```

File: ring-jetty-adapter/src/ring/adapter/jetty.clj [hunk 2]
```diff
@@ -150,10 +172,15 @@
 (defn- ^Server create-server [options]
   (let [pool   (or (:thread-pool options) (create-threadpool options))
         server (Server. pool)]
-    (when (:http? options true)
+    (when (:http? options (or ;; default is enabled when no socket specified
+                              (not (:unix-socket options))
+                              ;; enabled with socket when host/port set
+                              (some options [:host :port])))
```
Comment (weavejester) [inline ring-jetty-adapter/src/ring/adapter/jetty.clj:178]: This feels too complex, and isn't what the documentation says. I think let's keep it simple and just leave it defaulting to true. Let's add a note to the `:unix-socket` option instead to tell people they may want to set `:http?` to false.

File: ring-jetty-adapter/src/ring/adapter/jetty.clj [hunk 3]
```diff
@@ -170,10 +197,14 @@
...
                             (defaults to true)
   :daemon?                - use daemon threads (defaults to false)
-  :http?                  - listen on :port for HTTP traffic (defaults to true)
+  :http?                  - listen on :port for HTTP traffic (defaults to true
+                            unless :unix-socket is set)
   :ssl?                   - allow connections over HTTPS
   :ssl-port               - the SSL port to listen on (defaults to 443, implies
                             :ssl? is true)
+  :unix-socket            - File to be used as a Unix domain socket; will be
+                            passed to [io/file]. Ensure there is no file at
```
Comment (weavejester) [inline ring-jetty-adapter/src/ring/adapter/jetty.clj:206]: I don't think `[io/file]` should be a link, as there's nothing to link it to.

File: ring-jetty-adapter/test/ring/adapter/test/jetty.clj [hunk 4]
```diff
@@ -12,8 +12,12 @@
            [java.net ServerSocket ConnectException]
            [java.security KeyStore]
            [java.io SequenceInputStream ByteArrayInputStream InputStream
-                    IOException]
-           [org.apache.http MalformedChunkCodingException]))
+                    IOException
+                    File]
```
Comment (weavejester) [inline ring-jetty-adapter/test/ring/adapter/test/jetty.clj:16]: These two lines should be one.

File: ring-jetty-adapter/test/ring/adapter/test/jetty.clj [hunk 5]
```diff
@@ -98,6 +102,23 @@
                          "text/plain"))
         (is (= (:body response) "Hello World")))))
 
+  (testing "socket server"
+    (let [sock (File/createTempFile "ring-jetty-server-" ".sock")]
+      (.delete sock)
+      (with-server hello-world {:unix-socket sock}
+        (is (= (-> (HttpClient/create)
+                   (.remoteAddress
+                     (reify Supplier
```
Comment (weavejester) [inline ring-jetty-adapter/test/ring/adapter/test/jetty.clj:111]: Indentation looks off here.

File: ring-jetty-adapter/test/ring/adapter/test/jetty.clj [hunk 6]
```diff
@@ -98,6 +102,23 @@
...
+                       (get [_]
+                         (DomainSocketAddress. (.getAbsolutePath sock)))))
+                   .get (.uri "/") .responseContent .aggregate .asString .block)
+               "Hello World")
+            "able to serve basic request")
+        (is (thrown-with-msg?
+              clojure.lang.ExceptionInfo #"File already exists"
```
Comment (weavejester) [inline ring-jetty-adapter/test/ring/adapter/test/jetty.clj:118]: Indentation looks off. Please see Clojure style guide.

File: ring-jetty-adapter/src/ring/adapter/jetty.clj [hunk 7]
```diff
@@ -130,6 +133,25 @@
...
+  (when (->> (System/getProperty "os.name")
+             (re-find #"(?i)^windows"))
+    (throw (ex-info "Unix sockets not supported on windows"
+                    {:os (System/getProperty "os.name")
+                     :unix-socket unix-socket})))
+  (let [http-factory (HttpConnectionFactory. (http-config options))
+        socket (io/file unix-socket)]
+    (when (.exists socket)
+      (throw (ex-info "File already exists at socket path; should be deleted before starting server"
+                      {:unix-socket unix-socket})))
```
Comment (weavejester) [inline ring-jetty-adapter/src/ring/adapter/jetty.clj:147]: Let's factor this out as well:

```clojure
(defn- check-socket-file-does-not-exist [socket]
  (when (.exists socket)
    (throw (ex-info "Could not create socket file: file already exists"
                    {:socket-file (str socket)}))))
```

**Review by weavejester — CHANGES_REQUESTED** [review]
Thanks for your hard work on this, and apologies for taking a while to get around to reviewing it.

**Comment by agorgl** [comment]
Usage of the new APIs is documented here:
https://jetty.org/docs/jetty/11/programming-guide/server/http.html#connector
https://jetty.org/docs/jetty/11/programming-guide/client/http.html#transport-unix-domain

**Comment by weavejester** [comment]
Fixed by: df469736241b7ada658dc58e4b7d46e6f2003b30

**Pull request closed without merging**